### PR TITLE
refactor: Refactor string input check

### PR DESCRIPTION
**Pull request by yinggeh**
#### What does the PR do?
Move each individual check of string input in model backend repos to the bakend repo. 

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [x] Related issues are referenced.
- [x] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [x] Added [test plan](#test-plan) and verified test passes.
- [x] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [ ] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [x] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [x] refactor

#### Related PRs:
backend: https://github.com/triton-inference-server/backend/pull/101
tensorflow_backend: https://github.com/triton-inference-server/tensorflow_backend/pull/104
pytorch_backend: https://github.com/triton-inference-server/pytorch_backend/pull/136
onnxruntime_backend: https://github.com/triton-inference-server/onnxruntime_backend/pull/263
python_backend: https://github.com/triton-inference-server/python_backend/pull/370

#### Test plan:
- CI Pipeline ID:
17021222 and 17057045

#### Background
Initially just to add missing string input tensor checks to python_backend. Refactored because duplicate code from other model backends were found.

### CODE REVIEW

File: include/triton/backend/backend_common.h [hunk 1]
```diff
@@ -671,4 +672,23 @@ TRITONSERVER_Error* BufferAsTypedString(
...
+/// @param element_idx Returns the number of validated strings.
+/// @param set_string_tensor_cb Callback function which sets string input tensor
+/// depending on the backend platform.
+/// @param onnx_backend Whether the backend platform is ONNX runtime.
+/// @return a TRITONSERVER_Error indicating success or failure.
+TRITONSERVER_Error* ValidateStringBuffer(
+    const char* buffer, size_t buffer_byte_size,
+    const size_t expected_element_cnt, const char* input_name,
+    size_t* element_idx,
+    const std::function<void(size_t, const char*, const uint32_t)>&
```
Comment (yinggeh) [inline include/triton/backend/backend_common.h:690]: I don't like how this callback function is passed into `ValidateStringBuffer` in order to add strings from buffer into the string tensor. Open to suggestions.

File: src/backend_common.cc [hunk 2]
```diff
@@ -1372,4 +1372,74 @@ GetRequestId(TRITONBACKEND_Request* request)
...
+    remaining_bytes -= len;
+    (*element_idx)++;
+  }
+
+  // Special handling for ONNX runtime backend
+  if (onnx_backend && remaining_bytes > 0) {
+    *const_cast<char*>(buffer) = 0;
+  }
+
+  if (*element_idx != expected_element_cnt) {
```
Comment (yinggeh) [inline src/backend_common.cc:1433]: Here slightly different condition checks from original ```tensorflow.cc``` and ```libtorch.cc```. Removing ```(*response != nullptr)```.

I don't think we want to skip sending this error or skip ```FillStringTensor``` if ```RESPOND_AND_SET_NULL_IF_ERROR``` is called before ```SetStringInputTensor``` in ```tensorflow.cc``` and ```libtorch.cc```. I believe it was a mistake because other string format checks before this line do not require ```if (*response != nullptr)``` to send error or ```FillStringTensor```. Open to discussions.
  ```
// Code from tensorflow.cc and libtorch.cc
if ((*response != nullptr) && (element_idx != request_element_cnt)) {
    RESPOND_AND_SET_NULL_IF_ERROR(
        response, TRITONSERVER_ErrorNew(
                      TRITONSERVER_ERROR_INTERNAL,
                      std::string(
                          "expected " + std::to_string(request_element_cnt) +
                          " strings for inference input '" + name + "', got " +
                          std::to_string(element_idx))
                          .c_str()));
    FillStringTensor(
        tensor, tensor_offset + element_idx, request_element_cnt - element_idx);
  }
```

File: include/triton/backend/backend_common.h (hunk 1 above)
Comment (yinggeh) [reply include/triton/backend/backend_common.h]: Based on discussion with Tanmay, I have separated the logic of setting string tensor from validating string buffer.

File: src/backend_common.cc [hunk 3]
```diff
@@ -1372,4 +1372,64 @@ GetRequestId(TRITONBACKEND_Request* request)
   return std::string("[request id: ") + request_id + "] ";
 }
 
+TRITONSERVER_Error*
+ValidateStringBuffer(
+    const char* buffer, size_t buffer_byte_size,
+    const size_t expected_element_cnt, const char* input_name,
+    std::vector<std::pair<const char*, const uint32_t>>* str_list)
```
Comment (Tabrizian) [inline src/backend_common.cc:1379]: Do we need to document that the lifetime of `str_list` is dependent on the validity of `buffer` object?

File: include/triton/backend/backend_common.h [hunk 4]
```diff
@@ -671,4 +672,17 @@ TRITONSERVER_Error* BufferAsTypedString(
 /// \return a formatted string for logging the request ID.
 std::string GetRequestId(TRITONBACKEND_Request* request);
 
+/// Validate the contiguous string buffer with correct format.
```
Comment (tanmayv25) [inline include/triton/backend/backend_common.h:675]:
```suggestion
/// Validate the contiguous string buffer with correct format and parse string elements into list of pairs of memory address and length. Note the returned list of pairs points to valid memory as long as memory pointed by buffer remains allocated.
```

**Review by tanmayv25 — APPROVED** [review]

**Review by Tabrizian — DISMISSED** [review]

**Review by Tabrizian — APPROVED** [review]

**Review by tanmayv25 — APPROVED** [review]
